include/navigation/image_database.h

```diff
@@ -412,47 +479,118 @@ class ImageDatabase
     template<class... Ts>
     auto getRouteRecorder(Ts&&... extraFieldNames)
     {
-        return RouteRecorder<sizeof...(extraFieldNames)>{
+        return RouteRecorder<ImageFileWriter, sizeof...(extraFieldNames)>{
             *this, std::forward<Ts>(extraFieldNames)...
         };
     }
 
+    //! Start recording a route, saving images into video file
+    template<class... Ts>
+    auto getRouteVideoRecorder(const cv::Size &resolution, hertz_t fps,
+                               Ts&&... extraFieldNames)
+    {
+        m_Resolution = resolution;
+        m_FrameRate = fps;
+
+        RouteRecorder<VideoFileWriter, sizeof...(extraFieldNames)> recorder{
+            *this, std::forward<Ts>(extraFieldNames)...
+        };
+        recorder.getMetadataWriter() << "video_file" << recorder.getVideoFileName()
+                                     << "frame_rate" << m_FrameRate.value();
+
+        return recorder;
+    }
+
+    hertz_t getFrameRate() const;
+
     //! Get the resolution of saved images
     cv::Size getResolution() const;
 
     //! Check if this database has any saved metadata (yet)
     bool hasMetadata() const;
 
+    template<class Func>
+    size_t forEachImage(const Func &func, bool greyscale = true) const
```

Comment:
I'm not really that familiar with this bit of the code but might it be worth templating ``ImageDatabase`` on a ``FrameReader`` policy/strategy class like ``Recorder`` does rather that having a monolithic thing (or just derive off separate classes)?

Reply:
I think you're right. In fact, I think it'd make sense to split out the reading and writing functionality into separate classes too by essentially turning the ``Recorder`` types into ``ImageDatabaseWriter``s. I'll have a go at refactoring

Reply:
Hmm... The problem is that we don't know the database type at load time and we want to be able to load either a video-type or multi-file-type database transparently. We *could* do this by making ``forEachImage`` virtual in some base class, but then we wouldn't be able to have ``Func`` as a template parameter. We could pass in a ``std::function`` instead and I think in this case the performance hit wouldn't be too much of a problem. What do you think?

Reply:
Ahh ok, in that case, it's totally fine as is